### PR TITLE
use BCSR in sparse format of jaxoperators instead of BCOO

### PR DESCRIPTION
**Pull request by PhilipVinc**
BCOO should just be used to construct it, not to use it for matrix vector multiplications...

### CODE REVIEW

**Comment by gcarleo** [comment]
any speedup?

**Comment by wdphy16** [comment]
Yes I agree. Although in `MCState` we don't need to multiply a NetKet operator on a dense vector, when working with `FullSumState` it indeed speeds up in my use case.

By the way, in `FullSumState` we can put the sparse matrix - dense vector multiplication into JIT to further speed it up, at the cost of longer compiling time (specializing on the sparsity structure can take a long time) and higher memory usage, as in [my commit](https://github.com/wdphy16/netket/commit/73b7e9ecee8c5740578c00b0d9394bb38131edd9)

**Review by wdphy16 — APPROVED** [review]

File: netket/operator/_discrete_operator_jax.py [hunk 1]
```diff
@@ -245,11 +246,9 @@ def to_qobj(self):  # -> "qutip.Qobj"
         )
 
     def __matmul__(self, other):
-        if (
-            isinstance(other, np.ndarray)
-            or isinstance(other, jnp.ndarray)
-            or isinstance(other, JAXSparse)
-        ):
+        if isinstance(other, JAXSparse):
```
Comment (PhilipVinc) [inline netket/operator/_discrete_operator_jax.py:249]: @wdphy16 the only questionable thing is that jax sparse matrices do not support multiplication by another jax CSR sparse matrix, so I had to ensure that when one tries to do `jaxoperator @ sparsevector` we convert it to dense first.

This breaks current behaviour where, as we use BCOO, we get a sparse vector out. 

But I think that's not important? Hopefully one never does this kind of multiplication...

**Comment by PhilipVinc** [comment]
@wdphy16 how bad is compile time in that case?

File: netket/operator/_discrete_operator_jax.py (hunk 1 above)
Comment (wdphy16) [reply netket/operator/_discrete_operator_jax.py:249]: Ah it's a bit unexpected for me, but anyway JAX sparse matrices are experimental so we can make breaking changes

In real use cases, I guess we usually first do NetKet operator multiplications then convert it to a JAX sparse matrix

**Comment by wdphy16** [comment]
> @wdphy16 how bad is compile time in that case?

I don't remember exactly, but for large experiments it's much less than the actual computation time, and for small experiments I did not feel much more lag

**Pull request merged**